<a href="https://colab.research.google.com/github/amiralitalebi/CN6005_2321004/blob/main/Week5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Dense, Flatten, Dropout, BatchNormalization
from tensorflow.keras.datasets import mnist
from sklearn.metrics import confusion_matrix


In [ ]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

print(X_train.shape)
print(X_test.shape)


11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
(60000, 28, 28)
(10000, 28, 28)


In [ ]:
# normalise pixel values (0–255 → 0–1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# reshape for CNN: (samples, height, width, channels)
X_train = X_train.reshape(-1, 28, 28, 1)
X_test = X_test.reshape(-1, 28, 28, 1)


In [ ]:
model = Sequential()

model.add(Conv2D(32, kernel_size=3, activation='relu', input_shape=(28,28,1)))
model.add(BatchNormalization())
model.add(Conv2D(32, kernel_size=3, activation='relu'))
model.add(BatchNormalization())
model.add(Conv2D(32, kernel_size=5, strides=2, padding='same', activation='relu'))
model.add(BatchNormalization())
model.add(Dropout(0.3))

model.add(Conv2D(64, kernel_size=3, activation='relu'))
model.add(BatchNormalization())
model.add(Conv2D(64, kernel_size=3, activation='relu'))
model.add(BatchNormalization())
model.add(Conv2D(64, kernel_size=5, strides=2, padding='same', activation='relu'))
model.add(BatchNormalization())
model.add(Dropout(0.3))

model.add(Conv2D(128, kernel_size=4, activation='relu'))
model.add(BatchNormalization())
model.add(Flatten())
model.add(Dropout(0.3))

model.add(Dense(10, activation='softmax'))

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])


/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
history = model.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.2)


Epoch 1/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 218s 283ms/step - accuracy: 0.8853 - loss: 0.3732 - val_accuracy: 0.9827 - val_loss: 0.0613
Epoch 2/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 248s 265ms/step - accuracy: 0.9818 - loss: 0.0598 - val_accuracy: 0.9875 - val_loss: 0.0448
Epoch 3/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 210s 279ms/step - accuracy: 0.9851 - loss: 0.0460 - val_accuracy: 0.9888 - val_loss: 0.0364
Epoch 4/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 209s 279ms/step - accuracy: 0.9879 - loss: 0.0379 - val_accuracy: 0.9891 - val_loss: 0.0387
Epoch 5/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 200s 266ms/step - accuracy: 0.9896 - loss: 0.0344 - val_accuracy: 0.9913 - val_loss: 0.0323


In [ ]:
test_loss, test_acc = model.evaluate(X_test, y_test)
print("Original CNN Test Accuracy:", test_acc)


In [ ]:
model2 = Sequential()

model2.add(Conv2D(16, kernel_size=3, activation='relu', input_shape=(28,28,1)))
model2.add(MaxPooling2D())

model2.add(Conv2D(32, kernel_size=3, activation='relu'))
model2.add(MaxPooling2D())

model2.add(Flatten())
model2.add(Dense(64, activation='relu'))
model2.add(Dense(10, activation='softmax'))

model2.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

history2 = model2.fit(X_train, y_train, epochs=3, batch_size=64, validation_split=0.2)


In [ ]:
test_loss2, test_acc2 = model2.evaluate(X_test, y_test)
print("Modified CNN Accuracy:", test_acc2)


This week I learned how to use a CNN to classify handwritten digits from the MNIST dataset. At first, I thought the model was too complex, but when I ran the code in Google Colab, I started to understand how each layer helps the network learn features from the images. The original CNN achieved very high accuracy, close to 99%, which showed me how powerful convolution layers are when working with image data.

After that, I reduced the number of convolution layers to see how the accuracy changes. The accuracy dropped slightly to around 97–98%. This helped me understand that more convolution layers allow the model to learn more detailed patterns. I also realised that deep models are not always necessary, but they do give better performance for complex tasks. Overall, this week helped me build confidence in using CNNs and understanding how the network depth affects the final accuracy.